decisions/2024-04-15_decimal_pack_size.md

```diff
@@ -0,0 +1,131 @@
+- Backend should keep full stored value to maximum accuracy. These values can be smoothed over in some cases (e.g 0.9999999999999997 can be rounded up and treated as 1 in UI, in the backend it'll be that quantity).
+- Calculations will accrue inaccuracy. For individual lines this is usually tolerable but it does accumulate! In tenders we see it in monetary values: summing the "adjust cost" of several thousand lines can result in whole dollars of inaccuracy. Governments/Procurement agencies may have policies of acceptable level of calculation inaccuracy.
+  - It is worth highlighting that number of packs and our monetary fields are all already floating point type with these same problems, so we must already handle all the above issues regardless. **So it's not really a con if we have to do it anyway.** It just makes the surface area of floating point accuracy a smidgen higher.
+- recurring numbers create headaches of who truly has the "full" amount of something. E.g
+
+### Option 2 - Round pack size up to next whole number and adjust number of packs, monetary fields etc.
+
+We convert decimal pack sizes up to the next whole number.
+
+E.g.
+
+1 pack of 1L is repacked to 40 packs of 0.05L in mSupply desktop. The latter on integration in OMS:
+
+1. pack size: 0.05L/0.05 = 1L
+2. number of packs: 40packs\*0.05 = 1pack
+3. cost/sell price: 0.5/0.05 = $10
+
+The pharmacist feedback I have received on this is that it'd be confusing to users if they had a system telling them they had 1 of when they can see there are 40 smaller vessels on the shelf. For accounting, I have not asked a finance stakeholder yet but I presume this is very very bad form in regards of adjusting the number of what was physically ordered/received and the cost of each item ordered.
+
+A potential solution for this is to use pack variants. After converting back to 1L pack size, we use a variant of "50mL bottle" that converts all the properties back in the UI layer. Pack variants don't appear to quite support this - the assumption is the base unit is a small whole number (integer). In this case we'd need a pack variant of 0.05L, which would do the reverse of the calculations above. There is additionally an issue of when and how these pack variants are generated, as they are centrally controlled data.
+
+Going to the option 1 problem statement, we also have to solve this at integration still so no avoided in anyway:
+
+> There is potential for discrepancies to occur, e.g. store A repacks `1 * 1L` to `3 * 0.33L`. There are 0.01 packs of the original line... do we just zero it? If we issue all 3 to another store, they receive 0.99L. Do we round it up? What's the threshold?
+
+_Pros:_
+
+- Open mSupply can keep using integer for pack sizes, which feels good in a programming lens. Floats with their inaccuracy are genuinely not ideal!
```

Comment:
One other pro, which is something I've been thinking about lately, is reduced cognitive load due to reduction of use cases. I think that's a metric that can help assess the very objective 'simplicity'. (something is simpler if it leads to reduced set of use cases to consider)

Like in this case we can say that decimal pack size leads to a lot more use cases to think through (all of the presentation layer stuff, that's a long time in discussion, the potential of both pack size and number of packs being decimal and the effect on accuracy of the stored data).

Reply:
I agree with the philosophy, but I think I disagree that it's less cognitive load:

1. Pack size is int so OMS sync translation will munge any lines that have decimal pack size. This must also be remembered in API for any future 3rd party integration (takes a lot to explain physical 100x 750mL bottles are being recorded as 75x 1L). Support/PMs must train pack variants to be used to achieve veneer of a particular type of repack (partial pack size), rather than using the repack functionality they'd usually use for the task. What if the wrong pack variant is being displayed? What if the appropriate pack variant isn't in the system? We will still have cognitive load of f64 calcs/imprecision as all related fields for arithmetic with pack_size are f64. Do we need to make sure any such records if munged are always pushed back to the server immediately? If from a transfer, does it now mismatch the other half or do we need to record what pack size to convert _back_ to when pushing to central server? Do repack transactions with decimal values now appear to have made no repack because we undid the conversion? Who is burdened to explain this to users?

or 

2. Pack size is float so sync translation is direct. If a 3rd party system also uses f64 we're fine (or they're using a nested pack sort of thing so a complex mapping is needed regardless of solution). UX is the same as Legacy mSupply too and relatively straightforward (use repack). We do have to deal with f64 precision in calculations, which in most contexts we already have to do for several related fields (rust calculations require us coercing i32 to f64 to do math with any of those f64 fields anyway, so we really don't avoid any bullets by keeping i32 here). We already require an imprecision mitigation strategy because number of packs is f64. We'd have to apply this strategy if munging records from f64 to i32 anyway as we recalculate the other f64 fields. UI may have to round certain imprecision away which we have to address anyway given the other f64 fields, so would reuse the components.

Option 1 sounds like a lot more new and unique cognitive load introduced, while option 2 to me is a whole lot stuff we already have to consider?

Reply:
> munged

I would say **adjusted**, in my view **munged** is what they are now (with fractional pack_size), because item wasn't defined in base units correctly (base units should have been lowest dispensable unit, or something that can make sense in a pack size for lower dispensable unit), in this case in warehouse they would have pack_size = 1000 and when 750, respectively with base unit of 1ml). 

>  If from a transfer, does it now mismatch the other half or do we need to record what pack size

What do we do when we repack to 1 (all of mobile)

> Do we need to make sure any such records if munged are always pushed back to the server immediately?

Nope, but yes something to consider

> Do repack transactions with decimal values now appear to have made no repack because we undid the conversion?

They appear as split (if pack size didn't change form original)

> Who is burdened to explain this to users?

How do you explain to user that their floats don't add up ? (sorry that's an anti question, but the point is, there is still something potentially to explain, which one is easier to explain ¯\_(ツ)_/¯)

> Option 1 sounds like a lot more new and unique cognitive load introduced, while option 2 to me is a whole lot stuff we already have to consider?

I agree in a way, but I also had a point of what if both of the number of packs and pack_size are float, how do you deal with precision then (which i thought is different to "We already require an imprecision mitigation strategy")

Tbh we had similar discussion when fractional number of packs was introduced to omSupply, kind of got overpowered with the fact that mSupply supports it, and now that same argument (that we now support it in omSupply) is used for this, so have to concede to go with option 1 (even though at heart I think both number of pack and pack size should be integer and defined as a smallest dispensable unit, in the first place)